Implement support for Playnite

polson opened the pull request:
Parses Playnite library files and adds games/platforms to Pegasus

Currently supports launching executables, emulators, and Steam games.

See: https://github.com/mmatyas/pegasus-frontend/issues/600

polson commented:
```
make[5]: *** No rule to make target 'data/basic/game/Test Bros Something.zip', needed by 'release/qrc_data.cpp'.  Stop.
```

Derp.  What did I miss?

mmatyas commented:
Awesome work! Will see if I can review the code tonight. I saw you use the Steam call here too, I guess with #861 we could reduce the duplication there. Will review that one more time too.

> Derp. What did I miss?

I don't see `basic/game/Test Bros (JU) [!].zip` and `basic/game/Test Bros Something.zip` in the list of changed files. Seems these are checked more strictly on Windows.

Also what is `tests/backend/providers/playnite/target_wrapper.bat`? Some AppVeyor workaround?

polson commented:
Oh right it appears I didn't check in those zip files in Git, whoops!

The Playnite tests I pretty much copy-pasted from the LaunchBox tests, and that `.bat` file was there too.  I actually have no idea what it is, lol.  Sounds like I don't need it.

Let me know if you have any comments on code style or if I'm doing anything in a goofy way.  I'm rather new to C++ and Qt and still learning some things :)

mmatyas commented on tests/backend/providers/providers.pro:
```diff
@@ -8,4 +8,5 @@ SUBDIRS += \
     playtime \
 
-win32: SUBDIRS += launchbox
+win32: SUBDIRS += launchbox \
```
Style: Please break LaunchBox into a new line too.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,77 @@
+#include <QJsonObject>
+#include <QSettings>
+#include <QString>
+#include <array>
```
This doesn't seem to be in use anywhere.

mmatyas reviewed (changes requested):
Looks good! Don't be alarmed about the number of comments, I'm somewhat infamous for detailed reviews :) In general, I've found some parts where the style/readability could be improved, possible optimizations and a few missing error checks. I could imagine moving some code around at certain places (eg. to avoid passing emit as a callback), but in general the structure looks file. Of course, for points where you have a different technical opinion, feel free to disagree with me!

mmatyas commented:
Ok, with #868 merged, it is now possible to get the Steam call with `providers::find_steam_call()`.

polson commented:
Made some changes:
1. Moved more stuff into `PegasusProvider` and removed the progress callback
2. Added validation on fields after parsing
3. Updated to get the Steam call the new way
4. Lots of cleanup

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.cpp:
```diff
@@ -0,0 +1,173 @@
+        emu.id = emu_json.get_string(QStringLiteral("Id"));
+        emu.name = emu_json.get_string(QStringLiteral("name"));
+        auto profiles = emu_json.get_json_array(QStringLiteral("Profiles"));
+        for (const auto profile : profiles) {
```
You're possibly making a copy of each profile. You can use `const auto& profile`, just to make sure. Writing out the type for `profile` (and `profiles` above) would make it even more readable (they are just QJsonArray/Object/Value I guess?).

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.cpp:
```diff
@@ -0,0 +1,173 @@
+            emu_profile.working_dir = profile_json.get_string(QStringLiteral("WorkingDirectory"));
+            emu_profile.id = profile_json.get_string(QStringLiteral("Id"));
+            emu_profile.name = profile_json.get_string(QStringLiteral("Name"));
+            emu.profiles.insert({ emu_profile.id, emu_profile });
```
`emplace` (and `emplace_back` for vectors) is preferred over `insert`, for optimization.

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.cpp:
```diff
@@ -0,0 +1,173 @@
+    while (dir_it.hasNext()) {
+        PlayniteGame game;
+        const QString file_path = dir_it.next();
+        QJsonObject json_object = get_json_object_from_file(file_path);
```
It seems this isn't modified either, perfect for a nice `const`.

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.cpp:
```diff
@@ -0,0 +1,173 @@
+            emu_profile.name = profile_json.get_string(QStringLiteral("Name"));
+            emu.profiles.insert({ emu_profile.id, emu_profile });
+        }
+        emulator_map.insert({ emu.id, emu });
```
You might want to make sure at least the most important fields are valid, before storing anything for further use. (Same for the other `parse_` functions.)

polson replied:
I initially did all of the validation during parsing, but it became hard to read and split up validation into many different places.  That is why I let the parsing continue with empty values and then did all of the validation all in one place.

I don't think there's really much of a performance penalty for doing things this way, because by the time we have a QJsonObject we already have it backed by a QHash with all of the fields.  So each of the `get` calls is constant time.

But if you prefer I can do the validation at the same time as the parsing instead.

mmatyas replied:
Good point; I see `validate_game` for example dealing with some of it, so that's fine too. I usually prefer parse time, to avoid accidentally using them, and to keep less things in memory, but if it leads to hard to maintain code, the one place check is good too.

polson replied:
I'll short circuit in the cases where the QJSonObject is empty (probably means the file doesn't exist)

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.h:
```diff
@@ -0,0 +1,51 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+#include "utils/HashMap.h"
```
Very minor style: Leave a blank line around the pragma.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+
+    QTextDocument doc;
+    doc.setHtml(game_info.description);
+    const QString& description = doc.toPlainText();
```
https://doc.qt.io/qt-5/qtextdocument.html#toPlainText returns a newly allocated QString, no need for the reference here.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+    if (!game.releaseDate().isValid())
+        game.setReleaseDate(QDate::fromString(game_info.release_date, Qt::ISODate));
+
+    if (game_info.community_score >= 0.0f) {
```
What if it's over ~9000~ 100?

polson replied:
Clamped from 0 - 100

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+
+Provider& PlayniteProvider::run(SearchContext& sctx)
+{
+    const auto playnite_dir_path = [this] {
```
Write out the type here, it's easy to have accidents with lambdas.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+    std::vector<model::Game*> output;
+    PlayniteLaunchHelper launch_helper;
+
+    const float num_progress_updates = 5;
```
Is there a particular reason why this is exactly 5?

polson replied:
Just a number I made up that seemed like "not too many progress updates".

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+            Log::info(display_name(), LOGMSG("No install directory found for game %1, skipping...").arg(game.name));
+            return false;
+        }
+    } else if (game.play_action.type == "1") {
```
Hm, didn't you introduce an enum for this?

polson replied:
I did, though it's only converted when we create a `PlayniteGameLaunchInfo` object.

I think I will move the enum out and see if there's a nice way to convert it.  I also should probably parse this as an int and not a string.

I like to keep my JSON models 1:1 with the data so that's why I didn't convert this into enum in the `PlayniteGame`.  It makes it easier to debug parsing issues imo.  But I agree it would make more sense to convert to an enum here first before doing these checks

mmatyas reviewed (changes requested):
Ok, looking good! Here's round 2, I've mostly just found some optimization and style improvement possibilities.

polson commented on src/backend/providers/playnite/PlayniteLaunchHelper.cpp:
```diff
@@ -0,0 +1,97 @@
+        launch_cmd = build_launch_cmd(exec_path, args);
+    } else if (launch_type == PlayniteGameLaunchInfo::LaunchType::EMULATOR) {
+        const PlayniteEmulator emulator = components.emulators.at(play_action.emulator_id);
+        const PlayniteEmulator::EmulatorProfile emu_profile = emulator.profiles.at(play_action.emulator_profile_id);
```
I thought `at` returned a reference which meant it wouldn't be making a copy.  I think I misunderstood how that works.

mmatyas replied:
It does return a reference, but the types on the left side are regular, fully allocated objects; the compiler will have to resolve the reference for them. Of course, you can just fix it by marking the types with a `&`.

polson reviewed (commented):
Thanks for the thorough reviews!  I added some comments to the comments

I'll see if I can get the auto formatter in check and make those changes

mmatyas commented on src/backend/providers/playnite/PlayniteLaunchHelper.cpp:
```diff
@@ -0,0 +1,97 @@
+        const QString exec_path = get_steam_call();
+        const QString args = QStringLiteral("%1 %2").arg(play_action.path, play_action.arguments);
+        launch_cmd = build_launch_cmd(exec_path, args);
+    } else if (launch_type == PlayniteGameLaunchInfo::LaunchType::EMULATOR) {
```
Yeah, I've tried setting up clang-format several times, but found it way too rigid and/or buggy. It'd be nice to enforce *some* rules and not others, but as far as I know it doesn't support such operation.

mmatyas commented on src/backend/providers/playnite/PlayniteLaunchHelper.h:
```diff
@@ -0,0 +1,55 @@
+        const;
+
+private:
+    mutable QString m_steam_call;
```
PlayniteLaunchHelper is created exactly once during a search, I think it's fine. In the case of many Steam games, it would be even faster if the value doesn't have to be checked on every call.

mmatyas commented on src/backend/providers/playnite/PlayniteMetadataParser.cpp:
```diff
@@ -0,0 +1,173 @@
+    , json_ext_list(QStringList() << QStringLiteral("*.json"))
+{}
+
+PlayniteComponents PlayniteMetadataParser::parse_metdata() const
```
Typo in `metdata`.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,307 @@
+            continue;
+        }
+
+        // PC is handled by Pegasus default theme as "IBM".  We want "Windows" instead.
```
The Steam provider imports games into the "Steam" collection ([see here](https://github.com/mmatyas/pegasus-frontend/blob/382cae0e73c3fc857c36cbdc20e1e6487db812ce/src/backend/providers/steam/SteamProvider.cpp#L172)); I don't think I have PC hardcoded anywhere. The IBM logo actually comes from RetroPie, where it used to mean emulated PC games (DosBox, etc.), among the various other consoles.

If "PC" is the default that Playnite uses, I'm fine with the change, the current comment (*"By default, Playnite categorizes Windows games as "PC", which is handled by Pegasus default theme using the IBM logo. For a good out-of-the-box experience, this should be "Windows"*) sound good to me.

polson replied:
Sounds good.  Playnite also has a category for "IBM PC".  Not sure if DosBox games import like that, I haven't tried it.  

I should probably test with some more datasets and see if everything works okay

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,306 @@
+            return false;
+        }
+
+        const PlayniteEmulator emu = emu_it->second;
```
Found another copy here, and below.

polson replied:
Thanks, I'll do a search for these

polson commented on src/backend/providers/playnite/PlayniteLaunchHelper.cpp:
```diff
@@ -0,0 +1,97 @@
+    if (type.isEmpty()) {
+        return PlayniteGameLaunchInfo::LaunchType::INSTALLED;
+    }
+    if (type == "1") {
```
Changed to use an Enum here

polson commented:
Made some more review changes

mmatyas commented:
Great! Let's see...

mmatyas commented on src/backend/providers/playnite/PlayniteComponents.h:
```diff
@@ -1,5 +1,5 @@
 // Pegasus Frontend
-// Copyright (C) 2017-2020  M�ty�s Mustoha
+// Copyright (C) 2017-2020  M�ty�s Mustoha
```
Um, what happened here?

polson replied:
LOL yikes.  I'll look into it

mmatyas commented on src/backend/providers/playnite/PlayniteLaunchHelper.cpp:
```diff
@@ -30,21 +30,25 @@ QString build_launch_cmd(const QString& exec_path, const QString& args)
         return PlayniteGameLaunchInfo::LaunchType::EMULATOR;
     }
     return PlayniteGameLaunchInfo::LaunchType::UNKNOWN;
 }
```
Since `type` is now a number, you could also use a `switch`, to shave off some code size.

mmatyas commented on src/backend/providers/playnite/PlayniteLaunchHelper.h:
```diff
@@ -40,15 +38,14 @@ struct PlayniteGameLaunchInfo
 private:
-    mutable QString m_steam_call;
-    QString get_steam_call() const;
+    QString m_steam_call;
```
This can be a const now, it seems we don't modify it anywhere.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -229,7 +228,8 @@ bool PlayniteProvider::validate_game(const PlayniteGame& game, const PlayniteCom
     }
 
-    if (game.play_action.type.isEmpty()) {
+    const PlayniteGameLaunchInfo::LaunchType& launch_type = PlayniteLaunchHelper::get_launch_type_enum(game.play_action.type);
```
`PlayniteLaunchHelper::get_launch_type_enum` returns a plain type, no need for reference.

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.h:
```diff
@@ -23,9 +23,8 @@ namespace playnite {
-{
-    Q_OBJECT
+class PlayniteProvider : public Provider {
+Q_OBJECT
```
Indent the Q_OBJECT here, and it'll be perfect

mmatyas reviewed (changes requested):
Ok, just some minor things left.

polson commented:
Made the requested changes.  Thanks for reviewing so quickly

mmatyas commented on src/backend/providers/playnite/PlayniteProvider.cpp:
```diff
@@ -0,0 +1,307 @@
+// Pegasus Frontend
+// Copyright (C) 2017-2020  M�ty�s Mustoha
```
Found another encoding error.

polson replied:
Weird that find/replace didn't find that.  I wonder if it got re-encoded weird somehow.  Still don't know what's causing it

mmatyas reviewed (changes requested):
Almost there! Other than this one, I think it's ready for merge. If there's nothing else you'd like to change, could you squash all commits into one and rebase to master? (I can also do it tomorrow too, if you wish.)

polson commented:
I'll fix the remaining issues and squash commits in a bit, thanks

mmatyas commented:
Well, I was about to merge this... then noticed that your editor also put a Unicode Byte Order Marks into the file :sweat_smile: 

![image](https://user-images.githubusercontent.com/4354863/127870463-956a5a3d-e360-40a9-b07e-2c9bb7a9255e.png)

Only into PlayniteJsonHelper.cpp and .h, though. I'll fix this for you.

polson commented:
Yikes.  Sorry about that.  Thanks for taking care of it.
